{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dominant-shakespeare",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== No GPU Detected ===\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tmsincomb/anaconda3/envs/samtools/lib/python3.9/site-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py:381: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from DeepVCF import pathing\n",
    "from DeepVCF import pandas as pd\n",
    "from DeepVCF import alignment\n",
    "from DeepVCF import dwgsim\n",
    "from DeepVCF import bcftools_vcf\n",
    "from DeepVCF.core import DeepVCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "loose-terry",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacteria = [\n",
    "#     'c_difficile',  # !\n",
    "#     'e_coli',\n",
    "#     'klebsiella',  # !\n",
    "#     'listeria',\n",
    "#     'neisseria',\n",
    "#     'salmonella',  # !\n",
    "#     'staphylococcus',\n",
    "    'streptococcus',\n",
    "    'tb'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "incorporate-paste",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "streptococcus\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "DeepVCF: {'hom_alt': {'Sensitivity': 0.9833785617367707, 'PPV': 0.9948524365133836, 'Accuracy': 0.9929770657302754, 'F1': 0.9890822244967588}, 'het': {'Sensitivity': 0.9873133129918099, 'PPV': 0.9857303190636524, 'Accuracy': 0.9817708333333334, 'F1': 0.9865211810012837}}\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "DeepVCF: {'hom_alt': {'Sensitivity': 0.9682692307692308, 'PPV': 0.9843597262952102, 'Accuracy': 0.9837497236347557, 'F1': 0.9762481822588464}, 'het': {'Sensitivity': 0.964080695423979, 'PPV': 0.9850846321434557, 'Accuracy': 0.966547192353644, 'F1': 0.9744694960212201}}\n",
      "tb\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "DeepVCF: {'hom_alt': {'Sensitivity': 0.9861174675819985, 'PPV': 0.9975308641975309, 'Accuracy': 0.9945153518888719, 'F1': 0.9917913310318374}, 'het': {'Sensitivity': 0.9912630288166769, 'PPV': 0.9915670039865072, 'Accuracy': 0.9885848239311013, 'F1': 0.9914149931013337}}\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "=== Building Pileup From Scratch ===\n",
      "=== Pilup Complete ===\n",
      "=== Tensors Complete ===\n",
      "DeepVCF: {'hom_alt': {'Sensitivity': 0.9713600239916029, 'PPV': 0.9966153846153846, 'Accuracy': 0.9891658189216683, 'F1': 0.9838256511504291}, 'het': {'Sensitivity': 0.9801965230536659, 'PPV': 0.9923477196204469, 'Accuracy': 0.9817245557350566, 'F1': 0.9862346946535858}}\n"
     ]
    }
   ],
   "source": [
    "def populate_train(folder, prefix, ref, be_rate):\n",
    "    _ = dwgsim(\n",
    "        ref_file=ref,\n",
    "        output_folder=folder,\n",
    "        output_prefix=prefix,\n",
    "        verbose=False,\n",
    "        **{\n",
    "            '-r': .005,  # mutation rate :: need to get about 20K mutations for usable training,\n",
    "            '-e': be_rate,\n",
    "            '-E': be_rate,\n",
    "            '-C': 50,  # mean coverage\n",
    "            '-1': 150, # read length\n",
    "            '-2': 150, # read length\n",
    "        }\n",
    "    )\n",
    "    _ = alignment.bwa_mem(\n",
    "        reference=ref,\n",
    "        reads=[\n",
    "            folder / f'{prefix}.bwa.read1.fastq', \n",
    "            folder / f'{prefix}.bwa.read2.fastq'\n",
    "        ],\n",
    "        output_folder=folder,\n",
    "        output_prefix=prefix,\n",
    "        verbose=False,\n",
    "    )\n",
    "    deepvcf = DeepVCF()\n",
    "    deepvcf.train(\n",
    "        reference_file=ref,\n",
    "        alignment_file=folder/(prefix + '.mapped.bam'),\n",
    "        vcf_file=folder/f'{prefix}.mutations.vcf',\n",
    "        mimimum_alignment_coverage=.75,\n",
    "        minimum_coverage=30,\n",
    "        heterozygous_threshold=.25,\n",
    "        minimum_variant_radius=15,\n",
    "        save_pileup_to_destination=folder/f'{prefix}.pileup.npy',\n",
    "        # use_saved_pileup=folder/f'{prefix}.pileup.npy',\n",
    "        verbose=0,\n",
    "    )\n",
    "    return deepvcf\n",
    "\n",
    "def populate_test(folder, prefix, ref, be_rate, deepvcf):\n",
    "    _ = dwgsim(\n",
    "        ref_file=ref,\n",
    "        output_folder=folder,\n",
    "        output_prefix=prefix,\n",
    "        verbose=False,\n",
    "        **{\n",
    "            '-r': .005,  # mutation rate :: need to get about 20K mutations for usable training,\n",
    "            '-e': be_rate,\n",
    "            '-E': be_rate,\n",
    "            '-C': 50,  # mean coverage\n",
    "            '-1': 150, # read length\n",
    "            '-2': 150, # read length\n",
    "        }\n",
    "    )\n",
    "    _ = alignment.bwa_mem(\n",
    "        reference=ref,\n",
    "        reads=[\n",
    "            folder / f'{prefix}.bwa.read1.fastq', \n",
    "            folder / f'{prefix}.bwa.read2.fastq'\n",
    "        ],\n",
    "        output_folder=folder,\n",
    "        output_prefix=prefix,\n",
    "        verbose=False,\n",
    "    )\n",
    "    dcf = deepvcf.create_vcf(\n",
    "        reference_file=ref,\n",
    "        alignment_file=folder/(prefix + '.mapped.bam'),\n",
    "        save_pileup_to_destination=folder/f'{prefix}.pileup.npy',\n",
    "        # use_saved_pileup=folder/f'{prefix}.pileup.npy',\n",
    "        mimimum_alignment_coverage=.75,\n",
    "        minimum_coverage=30,\n",
    "        heterozygous_threshold=.25,\n",
    "        minimum_variant_radius=15,\n",
    "        output_folder=folder.parent,  # save as an actual .vcf file\n",
    "        output_prefix=prefix, \n",
    "    )\n",
    "    metrics = deepvcf.validation(\n",
    "        predicted_vcf=dcf,\n",
    "        real_vcf=folder/(prefix+'.mutations.vcf'),\n",
    "    )\n",
    "    print('DeepVCF:', metrics)\n",
    "    del deepvcf\n",
    "\n",
    "for name in bacteria:\n",
    "    print(name)\n",
    "    train_folder = pathing(name) / 'train'     \n",
    "    test_folder = pathing(name) / 'test'\n",
    "    ref_train = list(train_folder.glob('*.fasta'))[0]\n",
    "    ref_test = list(test_folder.glob('*.fasta'))[0]\n",
    "    for be, be_rate in [('2', .02), ('10', .1)]:\n",
    "        prefix=f'train-base-error-{be}percent'\n",
    "        deepvcf = populate_train(train_folder, prefix, ref_train, be_rate)\n",
    "        prefix=f'test-base-error-{be}percent'\n",
    "        populate_test(test_folder, prefix, ref_test, be_rate, deepvcf)\n",
    "            "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:samtools]",
   "language": "python",
   "name": "conda-env-samtools-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
